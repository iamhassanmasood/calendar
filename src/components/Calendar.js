import React, { useState, useEffect } from "react";
import moment from "moment";
import DateCard from "./DateCard";
import "antd/dist/antd.less";
import { Carousel, Row, Col, Typography, TimePicker } from "antd";
import {
  DownOutlined,
  UpOutlined,
  PlusCircleOutlined,
  MinusCircleOutlined,
  LeftOutlined,
  RightOutlined,
} from "@ant-design/icons";
import PropTypes from "prop-types";

const { Paragraph } = Typography;

const SamplePrevArrow = (props) => {
  const { className, style, onClick } = props;
  return (
    <LeftOutlined
      className={className}
      style={{
        ...style,
        display: "block",
        fontSize: "20px",
        color: "black",
      }}
      onClick={onClick}
    />
  );
};

const SampleNextArrow = (props) => {
  const { className, style, onClick } = props;
  return (
    <RightOutlined
      className={className}
      style={{
        ...style,
        display: "block",
        fontSize: "20px",
        color: "black",
      }}
      onClick={onClick}
    />
  );
};

function Calendar(props) {
  const [records, setRecords] = useState([]);
  const [selectedDate, setSelectedDate] = useState("Today");
  const [timeValue, setTimeValue] = useState("00:00");
  const [durationValue, setDurationValue] = useState("00:00");
  const [durationHours, setDurationHours] = useState(0);
  const [durationMinutes, setDurationMinutes] = useState(0);
  const [show, setShow] = useState(true);

  const generateDates = (props) => {
    const date = moment(props.firstDate);
    const disabledDates = props.disabledDates ? props.disabledDates : [];

    const first = props.firstDate
      ? moment(props.firstDate)
      : moment(new Date());
    const last = props.lastDate ? moment(props.lastDate) : null;

    const numberOfDays = last
      ? moment.duration(last.diff(first)).asDays() + 1
      : props.numberOfDays;

    const dates = [];
    for (let i = 0; i < numberOfDays; i += 1) {
      const isDisabled = !!disabledDates.includes(date.format("YYYY-MM-DD"));

      dates.push({
        date: `${date.format("D")}. ${date.format("MMMM")} ${date.format(
          "YYYY"
        )}`,
        day: date.format("D"),
        day_of_week: date.format("dddd"),
        month: date.format("MMMM"),
        disabled: isDisabled,
      });
      date.add(1, "days");
    }
    return setRecords(dates);
  };

  useEffect(() => {
    generateDates(props);
  }, []);

  const carouselSettings = {
    autoplay: props.autoplay,
    infinite: props.infinite,
    pauseOnHover: props.pauseOnHover,
    slidesToShow: props.slidesToShow,
    slidesToScroll: props.slidesToScroll,
    speed: props.speed,
    nextArrow: <SampleNextArrow />,
    prevArrow: <SamplePrevArrow />,
    responsive: [
      {
        breakpoint: 1360,
        settings: {
          slidesToShow: 3,
          slidesToScroll: 3,
        },
      },
      {
        breakpoint: 1024,
        settings: {
          slidesToShow: 2,
          slidesToScroll: 2,
        },
      },
      {
        breakpoint: 680,
        settings: {
          slidesToShow: 1,
          slidesToScroll: 1,
        },
      },
    ],
  };

  function handleTimeChange(time) {
    let _time = moment(time).format("HH:mm");
    setTimeValue(_time);
  }
  function handleDurationChange(time) {
    let _time = moment(time).format("HH:mm");
    setDurationValue(_time);
    let arr = _time.split(":");
    setDurationHours(Number(arr[0]));
    setDurationMinutes(Number(arr[1]));
  }

  function onIncreaseDuration() {
    if (durationMinutes < 59 && durationMinutes >= 0) {
      setDurationMinutes((x) => x + 1);
    } else if (durationMinutes === 59) {
      setDurationHours((x) => x + 1);
      setDurationMinutes(0);
    }
  }

  function onDecreaseDuration() {
    if (durationMinutes > 0) {
      setDurationMinutes((x) => x - 1);
    }
  }

  return (
    <Row>
      <Col span={12} offset={6}>
        <Row>
          <Col span={24}>
            <Paragraph className="date-link">Date</Paragraph>
            <Row className="date-hide-show">
              <Paragraph style={{ paddingRight: "10px", fontSize: "16px" }}>
                {selectedDate}
              </Paragraph>
              {show ? (
                <DownOutlined
                  onClick={() => setShow((x) => !x)}
                  style={{ fontSize: "20px" }}
                />
              ) : (
                <UpOutlined
                  onClick={() => setShow((x) => !x)}
                  style={{ fontSize: "20px" }}
                />
              )}
            </Row>
          </Col>
        </Row>
        <>
          {show ? (
            <Row>
              <Col span={24} className="carousal-div">
                <Carousel
                  autoplay
                  {...carouselSettings}
                  swipeToSlide
                  draggable
                  arrows={true}
                >
                  {records.map((item, idx) => (
                    <DateCard
                      month={item.month}
                      date={item.day}
                      day={item.day_of_week}
                      slected={item.date}
                      key={idx}
                      onChange={setSelectedDate}
                      disabled={item.disabled}
                    />
                  ))}
                </Carousel>
              </Col>
            </Row>
          ) : (
            <>
              <Row>
                <Col span={24}>
                  <Paragraph className="date-link">Time</Paragraph>
                  <Row className="date-hide-show">
                    <TimePicker
                      style={{ width: "60px" }}
                      defaultValue={moment(timeValue, "HH:mm")}
                      value={moment(timeValue, "HH:mm")}
                      format={"HH:mm"}
                      onChange={(time) => handleTimeChange(time)}
                      suffixIcon={null}
                      clearIcon={null}
                      bordered={false}
                    />
                    <DownOutlined style={{ fontSize: "20px" }} />
                  </Row>
                </Col>
              </Row>

              {timeValue !== "00:00" ? (
                <Row>
                  <Col span={24}>
                    <Paragraph className="date-link">Duration</Paragraph>
                    <Row className="date-hide-show">
                      <PlusCircleOutlined
                        onClick={onIncreaseDuration}
                        style={{ fontSize: "24px" }}
                      />
                      <TimePicker
                        style={{ width: "60px" }}
                        defaultValue={moment(durationValue, "HH:mm")}
                        value={moment(
                          `${durationHours}:${durationMinutes}`,
                          "HH:mm"
                        )}
                        format={"HH:mm"}
                        onChange={(time) => handleDurationChange(time)}
                        suffixIcon={null}
                        clearIcon={null}
                        bordered={false}
                      />
                      <MinusCircleOutlined
                        onClick={onDecreaseDuration}
                        style={{ fontSize: "24px" }}
                      />
                    </Row>
                  </Col>
                </Row>
              ) : (
                ""
              )}
            </>
          )}
        </>
      </Col>
    </Row>
  );
}

Calendar.propTypes = {
  firstDate: PropTypes.string.isRequired,
  lastDate: PropTypes.string,
  disabledDates: PropTypes.array,
  numberOfDays: PropTypes.number,
  autoPlay: PropTypes.bool,
  slidesToShow: PropTypes.number,
  slidesToScroll: PropTypes.number,
  speed: PropTypes.number,
  pauseOnHover: PropTypes.bool,
  infinite: PropTypes.bool,
  arrows: PropTypes.bool,
  onTimeChange: PropTypes.func,
};
const toDate = moment(Date()).format("YYYY-MM-DD");
Calendar.defaultProps = {
  firstDate: toDate,
  lastDate: "",
  disabledDates: [
    "2022-03-15",
    "2022-03-18",
    "2022-03-21",
    "2022-03-24",
    "2022-03-27",
  ],
  numberOfDays: 30,
  autoPlay: true,
  slidesToShow: 4,
  slidesToScroll: 1,
  speed: 300,
  pauseOnHover: true,
  infinite: false,
  arrows: false,
};

export default Calendar;
