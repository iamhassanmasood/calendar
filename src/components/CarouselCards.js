import React from "react";
import DateCard from "./DateCard";
import "antd/dist/antd.less";
import { Carousel, Row, Col } from "antd";
import { SampleNextArrow, SamplePrevArrow } from "./CarousalArrows";

export default function CarouselCards(props) {
  const carouselSettings = {
    autoplay: props.autoplay,
    infinite: props.infinite,
    pauseOnHover: props.pauseOnHover,
    slidesToShow: props.slidesToShow,
    slidesToScroll: props.slidesToScroll,
    speed: props.speed,
    nextArrow: <SampleNextArrow />,
    prevArrow: <SamplePrevArrow />,
    responsive: [
      {
        breakpoint: 1360,
        settings: {
          slidesToShow: 3,
          slidesToScroll: 3,
        },
      },
      {
        breakpoint: 1024,
        settings: {
          slidesToShow: 2,
          slidesToScroll: 2,
        },
      },
      {
        breakpoint: 680,
        settings: {
          slidesToShow: 1,
          slidesToScroll: 1,
        },
      },
    ],
  };
  return (
    <Row>
      <Col span={24} className="carousal-div">
        <Carousel
          autoplay
          {...carouselSettings}
          swipeToSlide
          draggable
          arrows={true}
        >
          {props.records.map((item, idx) => (
            <DateCard
              month={item.month}
              date={item.day}
              day={item.day_of_week}
              slected={item.date}
              key={idx}
              onChange={props.call}
              disabled={item.disabled}
            />
          ))}
        </Carousel>
      </Col>
    </Row>
  );
}
