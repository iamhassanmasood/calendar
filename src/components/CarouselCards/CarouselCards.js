import React from "react";
import DateCard from "../DateCard/DateCard";
import { Carousel, Col } from "antd";
import { dateFormater } from "../../helpers/formats";

export default function CarouselCards(props) {
  const carouselSettings = {
    autoplay: false,
    infinite: false,
    slidesToShow: props.datesToShow,
    slidesToScroll: props.steps,
    speed: props.speed,
    nextArrow: props.nextArrow,
    prevArrow: props.previousArrow,
    responsive: [
      {
        breakpoint: 1360,
        settings: {
          slidesToShow: 3,
          slidesToScroll: 3,
        },
      },
      {
        breakpoint: 1024,
        settings: {
          slidesToShow: 2,
          slidesToScroll: 2,
        },
      },
      {
        breakpoint: 680,
        settings: {
          slidesToShow: 1,
          slidesToScroll: 1,
        },
      },
    ],
  };

  return (
    <Col span={24} className="carousal-div">
      <Carousel
        autoplay={props.autoplay}
        swipeToSlide={props.swipeToSlide}
        arrows={props.arrows}
        draggable={false}
        dots={false}
        {...carouselSettings}
      >
        {props.records
          ? props.records.map((item, idx) => (
              <DateCard
                headStyles={props.cardHeaderStyle}
                bodyStyles={props.cardBodyStyle}
                headStylesClose={props.cardHeaderStyleClose}
                bodyStylesClose={props.cardBodyStyleClose}
                month={item.month}
                date={item.day}
                day={item.dayOfWeek}
                slectedDate={dateFormater(item.date)}
                key={idx}
                onChange={props.call}
                disabled={item.disabled}
              />
            ))
          : ""}
      </Carousel>
    </Col>
  );
}
