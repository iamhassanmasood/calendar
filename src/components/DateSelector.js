import React from "react";
import "antd/dist/antd.less";
import { Row, Col, Typography } from "antd";
import { DownOutlined, UpOutlined } from "@ant-design/icons";

const { Paragraph } = Typography;

export default function DateSelector({ show, selectedDate, onShow }) {
  return (
    <Row>
      <Col span={24}>
        <Paragraph className="date-link">Date</Paragraph>
        <Row className="date-hide-show">
          <Paragraph style={{ paddingRight: "10px", fontSize: "16px" }}>
            {selectedDate}
          </Paragraph>
          {show ? (
            <DownOutlined onClick={onShow} style={{ fontSize: "20px" }} />
          ) : (
            <UpOutlined onClick={onShow} style={{ fontSize: "20px" }} />
          )}
        </Row>
      </Col>
    </Row>
  );
}
