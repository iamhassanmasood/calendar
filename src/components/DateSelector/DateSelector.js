import React from "react";
import { Row, Col, Typography } from "antd";
import { DownOutlined, UpOutlined } from "@ant-design/icons";

const { Paragraph } = Typography;

export default function DateSelector({ show, selectedDate, onShow }) {
  return (
    <Row>
      <Col span={24}>
        <Paragraph className="date-link">Date</Paragraph>
        <Row className="date-hide-show">
          <Paragraph className="date-title-top">{selectedDate}</Paragraph>
          {show ? (
            <DownOutlined className="icon-class" onClick={onShow} />
          ) : (
            <UpOutlined className="icon-class" onClick={onShow} />
          )}
        </Row>
      </Col>
    </Row>
  );
}
