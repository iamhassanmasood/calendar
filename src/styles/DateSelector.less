.date-link {
  float: left;
  font-size: 16px;
  font-weight: normal;
  text-align: center;
  padding: 15px;
  color: #1890ff;
}
.date-hide-show {
  float: right;
  font-weight: normal;
  text-align: center;
  padding: 15px;
  color: #1890ff;
}
.date-title {
  padding-right: 10px;
  font-size: 16px;
  color: #000;
}
.icon-class {
  font-size: 20px;
  margin-top: 5px;
}
