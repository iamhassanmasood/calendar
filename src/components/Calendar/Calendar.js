import React, { useState, useEffect } from "react";
import moment from "moment";
import { Row, Col } from "antd";
import PropTypes from "prop-types";
import DateSelector from "../DateSelector/DateSelector";
import TimeSelector from "../TimeSelector/TimeSelector";
import DurationSelector from "../DurationSelector/DurationSelector";
import CarouselCards from "../CarouselCards/CarouselCards";

function Calendar(props) {
  const [records, setRecords] = useState([]);
  const [selectedDate, setSelectedDate] = useState("Today");
  const [timeValue, setTimeValue] = useState("00:00");
  const [durationValue, setDurationValue] = useState("00:00");
  const [durationHours, setDurationHours] = useState(0);
  const [durationMinutes, setDurationMinutes] = useState(0);
  const [show, setShow] = useState(true);

  const generateDates = (props) => {
    const firstDate = moment(props.firstDate);
    const disabledDates = props.disabledDates ? props.disabledDates : [];

    const first = props.firstDate
      ? moment(props.firstDate)
      : moment(new Date());
    const lastDate = props.lastDate ? moment(props.lastDate) : null;

    const numberOfDays = lastDate
      ? moment.duration(lastDate.diff(first)).asDays() + 1
      : props.numberOfDays;

    const dates = [];
    for (let i = 0; i < numberOfDays; i += 1) {
      const isDisabled = !!disabledDates.includes(
        firstDate.format("YYYY-MM-DD")
      );

      dates.push({
        date: `${firstDate.format("D")}. ${firstDate.format(
          "MMMM"
        )} ${firstDate.format("YYYY")}`,
        day: firstDate.format("D"),
        dayOfWeek: firstDate.format("dddd"),
        month: firstDate.format("MMMM"),
        disabled: isDisabled,
      });
      firstDate.add(1, "days");
    }
    return setRecords(dates);
  };

  useEffect(() => {
    generateDates(props);
  }, []);

  function handleTimeChange(time) {
    let _time = moment(time).format("HH:mm");
    setTimeValue(_time);
  }
  function handleDurationChange(time) {
    let _time = moment(time).format("HH:mm");
    setDurationValue(_time);
    let arr = _time.split(":");
    setDurationHours(Number(arr[0]));
    setDurationMinutes(Number(arr[1]));
  }

  function onIncreaseDuration() {
    if (durationMinutes < 59 && durationMinutes >= 0) {
      setDurationMinutes((x) => x + 1);
    } else if (durationMinutes === 59) {
      setDurationHours((x) => x + 1);
      setDurationMinutes(0);
    }
  }

  function onDecreaseDuration() {
    if (durationMinutes > 0) {
      setDurationMinutes((x) => x - 1);
    }
  }

  return (
    <Row>
      <Col span={12} offset={6}>
        <DateSelector
          show={show}
          onShow={() => setShow((x) => !x)}
          selectedDate={selectedDate}
        />
        <>
          {show ? (
            <CarouselCards
              {...props}
              records={records}
              call={setSelectedDate}
            />
          ) : (
            <>
              <TimeSelector
                value={moment(timeValue, "HH:mm")}
                defaultValue={moment(timeValue, "HH:mm")}
                format={"HH:mm"}
                handleChange={(time) => handleTimeChange(time)}
              />

              {timeValue !== "00:00" ? (
                <DurationSelector
                  handleIncrease={onIncreaseDuration}
                  handleDecrease={onDecreaseDuration}
                  handleChange={(time) => handleDurationChange(time)}
                  value={moment(`${durationHours}:${durationMinutes}`, "HH:mm")}
                  defaultValue={moment(durationValue, "HH:mm")}
                />
              ) : (
                ""
              )}
            </>
          )}
        </>
      </Col>
    </Row>
  );
}

Calendar.propTypes = {
  firstDate: PropTypes.string.isRequired,
  lastDate: PropTypes.string,
  disabledDates: PropTypes.array,
  numberOfDays: PropTypes.number,
  autoPlay: PropTypes.bool,
  slidesToShow: PropTypes.number,
  slidesToScroll: PropTypes.number,
  speed: PropTypes.number,
  pauseOnHover: PropTypes.bool,
  infinite: PropTypes.bool,
  arrows: PropTypes.bool,
};
const toDate = moment(Date()).format("YYYY-MM-DD");
Calendar.defaultProps = {
  firstDate: toDate,
  lastDate: "",
  disabledDates: [
    "2022-03-15",
    "2022-03-18",
    "2022-03-21",
    "2022-03-24",
    "2022-03-27",
  ],
  numberOfDays: 30,
  autoPlay: true,
  slidesToShow: 4,
  slidesToScroll: 1,
  speed: 100,
  pauseOnHover: true,
  infinite: false,
  arrows: true,
};

export default Calendar;
