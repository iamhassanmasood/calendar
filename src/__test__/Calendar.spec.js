import React from "react";
import { shallow } from "enzyme/build";
import Calendar from "./../components/Calendar/Calendar";
import DateSelector from "./../components/DateSelector/DateSelector";
import CarouselCards from "./../components/CarouselCards/CarouselCards";

describe("Calendar component", () => {
  it("calendar without crashing", () => {
    const wrapper = shallow(<Calendar />);
    expect(wrapper).toMatchSnapshot();
    wrapper.unmount();
  });
  it("check date selector props show it must be true", () => {
    const wrapper = shallow(<Calendar />);
    expect(wrapper.find(DateSelector).prop("show")).toBe(true);
  });
  it("check carousel cards props records property", () => {
    const wrapper = shallow(<Calendar />);
    expect(wrapper.find(CarouselCards).prop("records"));
  });
});
