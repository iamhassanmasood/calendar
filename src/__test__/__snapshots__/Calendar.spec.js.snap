// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar component calendar without crashing 1`] = `ShallowWrapper {}`;
